projects/royco/index.js

```diff
@@ -1,19 +1,140 @@
+  return `https://api.goldsky.com/api/public/${roycoSubgraph.projectId}/subgraphs/royco-recipe-${slug[chainId].royco}/${roycoSubgraph.recipe.version}/gn`;
+};
+
+const getVaultSubgraphUrl = (chainId) => {
+  return `https://api.goldsky.com/api/public/${roycoSubgraph.projectId}/subgraphs/royco-vault-${slug[chainId].royco}/${roycoSubgraph.vault.version}/gn`;
+};
 
 const config = {
-  ethereum: { factory: '0x75e502644284edf34421f9c355d75db79e343bca', fromBlock: 21244948 },
-  arbitrum: { factory: '0x75e502644284edf34421f9c355d75db79e343bca', fromBlock: 277208990 },
-  base: { factory: '0x75e502644284edf34421f9c355d75db79e343bca', fromBlock: 22754606 },
-}
+  [slug[1].defillama]: {
+    chainId: 1,
+    recipeSubgraphUrl: getRecipeSubgraphUrl(1),
+    vaultSubgraphUrl: getVaultSubgraphUrl(1),
+  },
+  [slug[42161].defillama]: {
+    chainId: 42161,
+    recipeSubgraphUrl: getRecipeSubgraphUrl(42161),
+    vaultSubgraphUrl: getVaultSubgraphUrl(42161),
+  },
+  [slug[8453].defillama]: {
+    chainId: 8453,
+    recipeSubgraphUrl: getRecipeSubgraphUrl(8453),
+    vaultSubgraphUrl: getVaultSubgraphUrl(8453),
+  },
+};
 
-Object.keys(config).forEach(chain => {
-  const { factory, fromBlock } = config[chain]
+Object.keys(config).forEach((chain) => {
+  const { recipeSubgraphUrl, vaultSubgraphUrl } = config[chain];
   module.exports[chain] = {
     tvl: async (api) => {
-      const logs = await getLogs2({ api, factory, eventAbi: 'event WrappedVaultCreated (address indexed underlyingVaultAddress, address indexed incentivizedVaultAddress, address owner, address inputToken, uint256 frontendFee, string name, string vaultSymbol)', fromBlock, })
-      const tokensAndOwners = logs.map(log => [log.underlyingVaultAddress, log.incentivizedVaultAddress])
-      return api.sumTokens({ tokensAndOwners })
+      const vaultMarketsQuery = gql`
+        {
+          rawMarkets {
+            inputTokenId
+            quantityOffered
+            incentivesOfferedIds
+            incentivesOfferedAmount
+            endTimestamps
+          }
+        }
+      `;
+
+      const vaultMarkets = await request(vaultSubgraphUrl, vaultMarketsQuery);
+
+      vaultMarkets.rawMarkets.map((market) => {
+        const inputTokenId = market.inputTokenId.split("-")[1];
+        const inputTokenAmount = market.quantityOffered;
+
+        api.add(inputTokenId, inputTokenAmount);
+
+        market.incentivesOfferedIds.forEach((rawIncentiveTokenId, index) => {
+          const incentiveTokenId = rawIncentiveTokenId.split("-")[1];
+          const incentiveTokenAmount = market.incentivesOfferedAmount[index];
+
+          const endTimestampInSeconds = new BigNumber(
+            market.endTimestamps[index]
+          );
+          const currTimestampInSeconds = new BigNumber(
+            Math.floor(Date.now() / 1000)
+          );
+
+          if (currTimestampInSeconds.isLessThan(endTimestampInSeconds)) {
+            api.add(incentiveTokenId, incentiveTokenAmount);
```

Comment:
can we read balances from the chain using an rpc instead? propietary apis are fine for reading config data

Reply:
It is not possible to read balances directly from chain data because of following limitations:

1. For Vault Markets:
- The wrapped vault is a  wrapper around another ERC4626 vault and therefore, wrapped vault itself holds no assets
- The previously committed version of royco TVL was counting balances of issued vault tokens of Wrapped Vault, rather than the assets deposited by the wrapped vault into the underlying ERC4626 vault -- this was making significant amount of TVL being excluded in summation

2. For Recipe Markets:
- The assets are deposited & withdrawn by the weiroll script (this is an arbitrary sequence of transactions which will differ for each market, like swap into some other token, then deposit half in one contract, rest in another, etc.) -- so we can't deterministically find the token address and it's token amount
- The amount deposited and received is not being stored on-chain, and is only emitted as an event -- these events are then indexed by subgraph, to calculate what amount is currently present inside the market
- The events are interdependent on one another, for example: in order to find out how much an LP deposited into the market, first we need to find the event that LP filled -> this gives the address of weiroll wallet (this wallet is generated uniquely every time an LP fills a position) -> then we need to find the offer id which was filled -> based on offer id, we need to find the IP event that had made the offer, this gives us the corresponding market id -> based on market id, we need to find the Market Creation event, this gives us the input asset that was deposited -> finally, this gives us liquidity of one one position of an LP -> Now, we need to find all positions of all LPs inside all markets -> therefore, this requires a cumulative calculation, rather than on-demand -- so we are doing them in subgraph.

Moreover, the entire code of subgraph is open-source and can be referenced in this repo: https://github.com/roycoprotocol/royco-subgraphs-template